### ISSUE TITLE
Questions regarding GK6X project

### ISSUE THREAD

**Issue by DonaldTsang** [pixeltris/GK6X#12]
I am considering the idea of buying GK61 for myself, and I have seen the GK ads stating that the keyboard mapping is stored on the keyboard IF using the Windows application. Is that true for GK6X as well?

**Comment by pixeltris** [comment]
Once the keys are mapped they stay on the keyboard. The official software has an additional layer which always requires constant communication with the keyboard, but there are 3 layers which can be mapped using the software and can be used forever on any device you connect your keyboard to (regardless of OS),

I imagine what they mean by "IF using the Windows application" is that the official software is a Windows application, so you can't map the keys without Windows. The code provided in this project is essentially a re-implementation of the official software which can run on other operating systems.

**Comment by DonaldTsang** [comment]
That is super sweet, I would most definitely use this along side my GK61. The RK61's arrow key bug is really killing me!

**Comment by pixeltris** [comment]
Hadn't heard of the RK61 before. Make sure to read up on any limitations for the GK61 before buying one. A keyboard which uses QMK is probably a wisest choice. I wanted a 75% hotswappable keyboard so my only real choice was GK84, I would have chosen something with a configurable firmware if such a keyboard existed.

I roughly know how the additional layer works but lighting data needs to be sent to the keyboard at a high throughput rate which can eat CPU usage so I didn't implement it. That layer has one nice trait in that macros are done via a special callback packet sent from the keyboard to the software, allowing for anything to be run when pressing a key. A drawback of that though is that macro key strokes then need to be implemented as software inputs rather than real hardware ones (which is another reason why I didn't implement it).

I'll possibly add that information to the readme at some point, but I imagine most people would have already got their keyboard by the time they look at this github, so that information would probably be a bit late by then!

**Comment by DonaldTsang** [comment]
> A keyboard which uses QMK is probably a wisest choice.

Which the whole GK series isn't, and also not RK61, CK62, CK61, DK61, Ajazz 61 and Maibenben 61.
Other 61 key models are just too expensive for me to handle.

> I roughly know how the additional layer works but lighting data needs to be sent to the keyboard at a high throughput rate which can eat CPU usage so I didn't implement it. 

Make sure to note why you are not implementing it. It is a good reason.

> I'll possibly add that information to the readme at some point, but I imagine most people would have already got their keyboard by the time they look at this github, so that information would probably be a bit late by then!

Caveat Emptor.

**Comment by pixeltris** [comment]
I'd recommend checking if those other keyboards support reprogramming the base layer fn keys. GK6X keyboards can be reprogrammed on the base layer, but not the keys on the base layer when the fn key is held down. The workaround that people use is a [key which switches to a different layer whilst it's pressed down](https://github.com/pixeltris/GK6X/blob/ae4f9e22265a914c2e52752cb4bf39a35f9252b2/GK6X/KeyValues.cs#L455), effectively acting as a separate fn key (also note that the position of the fn key itself cannot be changed).

**Comment by DonaldTsang** [comment]
I recommend trying the original software to see if it supports this Fn programming feature in the first place.

**Comment by DonaldTsang** [comment]
This might be important... http://jkdz.yunshubiao.com/thread-13465-1-1.html it states that the Fn layer is not programmable, you can only program the Fn key...

**Comment by pixeltris** [comment]
Yes I used the official software to intercept and document the messages the official software sends to the keyboard, which I then used to create this project.

You can map keys on the base layer, and layers 1-3. You can't map keys on the base layer's fn key set. You can map keys on layers 1-3 fn key set. You can't change the position of the fn key, but you can define a new key which is able to switch to a separate layer which people tend to use as an alternative to the fn key.

What I described might be a little bit hard to follow, but if you ever get one of the GK6X range keyboards it will become clear what that means.

**Comment by DonaldTsang** [comment]
No, I understand that correctly 100%, in that case it is out of scope to do something outside of the software, and just try and wrap this program around QMK or some other professional configurator.

**Comment by ondvavr** [comment]
> You can map keys on the base layer, and layers 1-3. You can't map keys on the base layer's fn key set. You can map keys on layers 1-3 fn key set. You can't change the position of the fn key, but you can define a new key which is able to switch to a separate layer which people tend to use as an alternative to the fn key.
> 
Could you please describe how the mapping configuration would look like to setup CapsLock or LShift to trigger the separate layer?

**Comment by AlainCh2** [comment]
Received today an MK14 (Magic Refiner) that should be a GK68,
 and landed here.
Need to set the Pause as Delete and some keys with specific color.

No idea how to use this program:
when launched it open a DOS window, and nothing happens.

What am I doing wrong?

Thanks
Alain
![2020-03-01 20 52 51](https://user-images.githubusercontent.com/41082160/75632666-aec22000-5bfe-11ea-8272-5582024b2fc0.jpg)

**Comment by kobusestas** [comment]
> Received today an MK14 (Magic Refiner) that should be a GK68,
> and landed here.
> Need to set the Pause as Delete and some keys with specific color.
> 
> No idea how to use this program:
> when launched it open a DOS window, and nothing happens.
> 
> What am I doing wrong?
> 
> Thanks
> Alain
> ![2020-03-01 20 52 51](https://user-images.githubusercontent.com/41082160/75632666-aec22000-5bfe-11ea-8272-5582024b2fc0.jpg)

I'm in a similar boat as AlainCh2, I'm not 100% sure how to run this program either, on Mac or Windows.

Any help would be appreciated!

**Comment by pixeltris** [comment]
I would suggest using the official software if you're on Windows. If it doesn't work with the official software, then this project wont work on it either as this project uses the data files from the official software.

You could possibly place a breakpoint down in the handshake code to see if it even detects the device.
